activesupport/test/cache/behaviors/cache_store_behavior.rb

```diff
@@ -498,7 +498,41 @@ def test_cache_miss_instrumentation
     ActiveSupport::Notifications.unsubscribe "cache_read.active_support"
   end
 
+  def test_nil_is_significant
+    skip "does not support blank key" if does_not_support_blank_keys?
+
+    @cache.write(nil, "baz")
+    assert_equal "baz", @cache.read(nil)
+
+    @cache.write([nil], "bizz")
+    assert_equal "bizz", @cache.read(nil)
+  end
+
+  def test_nil_is_significant_in_array_and_hash
+    @cache.write([ :a, nil, :b, "", :c ], "foo")
+    assert_equal "foo", @cache.read("a//b//c")
+
+    @cache.write({ hello: nil, world: "today", nil: "rules" }, "bar")
+    assert_equal "bar", @cache.read("hello=/nil=rules/world=today")
```

Comment:
Can cover these cases in the hash test above as well.

Reply:
I'll also move this testing logic over to the Key class.